src/Sentry/MetricAggregator.cs

```diff
@@ -0,0 +1,455 @@
+            foreach (var key in Buckets.Keys)
+            {
+                var bucketTime = DateTimeOffset.FromUnixTimeSeconds(key);
+                if (bucketTime < cutoff)
+                {
+                    yield return key;
+                }
+            }
+        }
+    }
+
+    private Dictionary<long, Dictionary<MetricResourceIdentifier, SentryStackFrame>> FlushableLocations()
+    {
+        _codeLocationLock.Wait();
+        try
+        {
+            var result = _pendingLocations;
+            _pendingLocations = new Dictionary<long, Dictionary<MetricResourceIdentifier, SentryStackFrame>>();
+            return result;
+        }
+        finally
+        {
+            _codeLocationLock.Release();
+        }
+    }
+
+    /// <summary>
+    /// Clear out stale seen locations once a day
+    /// </summary>
+    private void ClearStaleLocations()
+    {
+        var now = DateTime.UtcNow;
+        var today = now.GetDayBucketKey();
+        if (lastClearedStaleLocations == today)
+        {
+            return;
+        }
+        // Allow 60 seconds for all code locations to be sent at the transition from one day to the next
+        const int staleGraceInMinutes = 1;
+        if (now.Minute < staleGraceInMinutes)
+        {
+            return;
+        }
+
+        foreach (var dailyValues in _seenLocations.Keys.ToArray())
+        {
+            if (dailyValues < today)
+            {
+                _seenLocations.TryRemove(dailyValues, out _);
+            }
+        }
+        lastClearedStaleLocations = today;
+    }
+
+    /// <inheritdoc cref="IAsyncDisposable.DisposeAsync"/>
+    public async ValueTask DisposeAsync()
+    {
+        _options.LogDebug("Disposing MetricAggregator.");
+
+        if (_disposed)
+        {
+            _options.LogDebug("Already disposed MetricAggregator.");
+            return;
+        }
+
+        _disposed = true;
+
+        try
+        {
+            // Request the LoopTask stop.
+            await _shutdownSource.CancelAsync().ConfigureAwait(false);
+
+            // Now wait for the Loop to stop.
+            // NOTE: While non-intuitive, do not pass a timeout or cancellation token here.  We are waiting for
+            // the _continuation_ of the method, not its _execution_.  If we stop waiting prematurely, this may cause
+            // unexpected behavior in client applications.
+            LoopTask.Wait();
+        }
+        catch (OperationCanceledException)
+        {
+            _options.LogDebug("Stopping the Metric Aggregator due to a cancellation.");
+        }
+        catch (Exception exception)
+        {
+            _options.LogError(exception, "Stopping the Metric Aggregator threw an exception.");
+        }
+        finally
+        {
+            _flushLock.Dispose();
+            _shutdownSource.Dispose();
+            LoopTask.Dispose();
+        }
+    }
+
+    public void Dispose()
+    {
+        DisposeAsync().GetAwaiter().GetResult();
```

Comment:
if the task throws (e.g cancelled) it'll throw. Best handle here (dispose shouldn't throw)

Also, should we just not have a sync dipose then>? Go only async

Reply:
I did try going Async... .NET Framework doesn't yet have the `IAsyncDisposable` interface so was complaining about missinc the Bcl.Interfaces package or something. It is just an interface so we could potentially conditionally shim it, but it seemed like a lot of complexity for very little gain, so I rolled it all back.

I'll try/catch the above though.